src/backend/distributed/commands/role.c

```diff
@@ -309,144 +417,118 @@ MakeVariableSetStmt(const char *config)
+	/* load valid unitl data from the heap tuple, use default of infinity if not set */
+	Datum rolValidUntilDatum = SysCacheGetAttr(AUTHNAME, tuple,
+											   Anum_pg_authid_rolvaliduntil, &isNull);
+	char *rolValidUntil = "infinity";
 	if (!isNull)
 	{
 		rolValidUntil = pstrdup((char *) timestamptz_to_str(rolValidUntilDatum));
 	}
 
-	stmt->options = lappend(stmt->options, makeDefElem("validUntil",
-													   (Node *) makeString(rolValidUntil),
-													   -1));
+	Node *validUntilStringNode = (Node *) makeString(rolValidUntil);
+	DefElem *validUntilOption = makeDefElem("validUntil", validUntilStringNode, -1);
+	options = lappend(options, validUntilOption);
 
-	return CreateAlterRoleIfExistsCommand(stmt);
+	return options;
 }
 
 
 /*
- * GenerateAlterRoleIfExistsCommandAllRoles creates ALTER ROLE commands
- * that copies all roles from the pg_authid table.
+ * GenerateCreateOrAlterRoleCommand generates ALTER ROLE command that copies a role from
+ * the pg_authid table.
  */
 List *
-GenerateAlterRoleIfExistsCommandAllRoles()
+GenerateCreateOrAlterRoleCommand(Oid roleOid)
 {
-	Relation pgAuthId = heap_open(AuthIdRelationId, AccessShareLock);
-	TupleDesc pgAuthIdDescription = RelationGetDescr(pgAuthId);
-	HeapTuple tuple = NULL;
-	List *commands = NIL;
-	const char *alterRoleQuery = NULL;
+	HeapTuple roleTuple = SearchSysCache1(AUTHOID, ObjectIdGetDatum(roleOid));
+	Form_pg_authid role = ((Form_pg_authid) GETSTRUCT(roleTuple));
 
-#if PG_VERSION_NUM >= PG_VERSION_12
-	TableScanDesc scan = table_beginscan_catalog(pgAuthId, 0, NULL);
-#else
-	HeapScanDesc scan = heap_beginscan_catalog(pgAuthId, 0, NULL);
-#endif
+	CreateRoleStmt *createRoleStmt = NULL;
+	AlterRoleStmt *alterRoleStmt = NULL;
+	if (EnableAlterRolePropagation)
+	{
+		alterRoleStmt = makeNode(AlterRoleStmt);
+		alterRoleStmt->role = makeNode(RoleSpec);
+		alterRoleStmt->role->roletype = ROLESPEC_CSTRING;
+		alterRoleStmt->role->location = -1;
+		alterRoleStmt->role->rolename = pstrdup(NameStr(role->rolname));
+		alterRoleStmt->action = 1;
+		alterRoleStmt->options = GenerateRoleOptionsList(roleTuple);
+	}
 
-	while ((tuple = heap_getnext(scan, ForwardScanDirection)) != NULL)
+	ReleaseSysCache(roleTuple);
+
+	List *completeRoleList = NIL;
+	if (createRoleStmt != NULL || alterRoleStmt != NULL)
 	{
-		const char *rolename = NameStr(((Form_pg_authid) GETSTRUCT(tuple))->rolname);
+		/* add a worker_create_or_alter_role command if any of them are set */
+		char *createOrAlterRoleQuery = CreateCreateOrAlterRoleCommand(
+			pstrdup(NameStr(role->rolname)),
+			createRoleStmt,
+			alterRoleStmt);
 
-		/*
-		 * The default roles are skipped, because reserved roles
-		 * cannot be altered.
-		 */
-		if (IsReservedName(rolename))
-		{
-			continue;
-		}
-		alterRoleQuery = GenerateAlterRoleIfExistsCommand(tuple, pgAuthIdDescription);
-		commands = lappend(commands, (void *) alterRoleQuery);
+		completeRoleList = lappend(completeRoleList, createOrAlterRoleQuery);
 	}
 
```

Comment:
The grant role code is omitted now. Is that intentional? We might still want to grant extra roles to superuser.

Reply:
grants are not supported in community, it used to be always `NULL` here.